### DISCUSSION TITLE
Zenoh-pico: Using it in PEER mode

### DISCUSSION THREAD

**Discussion by nagygabor93** [eclipse-zenoh/roadmap#89] [Q&A] [answered]
I successfully used Zenoh-pico in client mode to stream sensor data to a zenohd router.

After that, I wanted to use peer mode, but for some reason it is not working, can not open a zenoh session. 
I have looked into the files and debugged a little bit, and figured out that this line is generating the error code:

In udp.c function: _z_endpoint_udp_multicast_valid:
const char *iface = _z_str_intmap_get(&endpoint->_config, UDP_CONFIG_IFACE_KEY);

The iface is NULL so the validity check function returns with _Z_ERR_CONFIG_LOCATOR_INVALID. 

Unfortunately, I could not find any more infos, and also, this check is only called one time in the stack, so no other examples.

I also ran a scouting before oppening the session, and for that I can see the router. I even tried it running on UDP since I figured it out that in zenoh-pico for PEER mode only UDP multicast is implemented currently (Am I righ here?)

Thanks for help in advance.

**Comment by Mallets** [comment]
Hello @nagygabor93 ,

Peer-to-peer  mode between Zenoh Pico and a Zenoh Router is not supported right now.
You can though use peer-to-peer mode in Zenoh Pico between Zenoh Pico applications via UDP multicast.
In other words, UDP multicast is not supported on the Zenoh Router for the time being (we are working on that).

**Reply by nagygabor93** [reply]
Hi, thanks for the answer, but I think my Zenoh-pico app is not reaching the point to start to try establish connection to the router. As I wrote, it  fails in _z_listen_link when it is checking the udp endpoint with _z_endpoint_udp_multicast_valid.

**Comment by cguimaraes** [comment, marked as answer]
Hello @nagygabor93,

Peer-to-peer mode in unicast (as supported by Zenoh Router) is not supported.
However, you can have peer-to-peer mode in Zenoh Pico between Zenoh Pico applications via e.g. UDP multicast.

For that, you need to change two configurations:
```C
    const char *mode = "peer";
    char *locator = "udp/224.0.0.225:7447#iface=eth0";

    z_owned_config_t config = z_config_default();
    zp_config_insert(z_loan(config), Z_CONFIG_MODE_KEY, z_string_make(mode));
    zp_config_insert(z_loan(config), Z_CONFIG_PEER_KEY, z_string_make(locator));
```

Note that the locator has the `iface` parameter, which is **mandatory** for Zenoh multicast transports over UDP multicast.
The fact that you are not setting it as part of your locator is the reason to obtain an invalid udp multicast locator.

**Reply by nagygabor93** [reply]
I forgot to mention that I am on ESP32, I have ported the cmake file to ESP32-IDF.

now I tried it whit setting "udp/224.0.0.224:7447#iface=eth0" and I am running a zenohd router on Ubuntu. 

The Zenoh-pico app now could open a session and streaming sensor data to somewhere without a problem :D It is configured in peer mode.

My problem is: I am not understanding completely the two modes: peer and client in zenoh-pico. I grasp something like that the client mode is opening a link between another endpoint (like a router) and streeming data there, and the peer is like don't really care about if someone is there or not?

Is there a documentation about the different modes somewhere (I could not find it, but maybe there is)? 

Thanks for the help!

**Reply by cguimaraes** [reply]
In the peer mode, nodes might actively look for neighbor nodes and establish peer-to-peer connections. The client mode always needs to be brokered by either a Zenoh Router or Peer.

For the moment, I think [here](https://drive.google.com/file/d/1673XJ03rdNB-a-JCuWhREMAT2bIUuVcg/view) is where you can find many of the Zenoh concepts explained. We also have a webinar series that might be interesting to grasp some of the concepts [here](https://www.youtube.com/watch?v=ahO4kT_Zg7s&list=PLZDEtJusUvAY04pwmpY8uqCG5iQ7NgSrR&pp=iAQB).
